kernel/sched.c

```diff
@@ -562,20 +562,16 @@ static void pend(struct k_thread *thread, _wait_q_t *wait_q, s32_t timeout)
 		}
 	}
 
-	if (timeout != K_FOREVER) {
-		s32_t ticks;
-
-		if (timeout < 0) {
-			timeout = 0;
-		}
-
-		ticks = _TICK_ALIGN + k_ms_to_ticks_ceil32(timeout);
+	if (!K_TIMEOUT_EQ(timeout, K_FOREVER)) {
+#ifdef CONFIG_LEGACY_TIMEOUT_API
+		timeout = _TICK_ALIGN + k_ms_to_ticks_ceil32(timeout);
+#endif
 
-		z_add_thread_timeout(thread, ticks);
+		z_add_thread_timeout(thread, timeout);
```

Comment:
there is some inconsistency there: `z_add_thread_timeout` takes `k_timeout_t` as the second argument. In legacy mode that means milliseconds and yet it seems that it takes ticks since ms to ticks conversion occurs.

Why not do that conversion for legacy mode in `z_add_timeout`?

Reply:
Mostly because I'm trying not to change the legacy mode code paths too much -- the code used to be done in the various subsystems above this function, so I left it that way.  The new code does, indeed, fold all the conversion code to a single place (the timeout constructor macros) and all the tick alignment to z_add_timeout().

Reply:
legacy will stay for long maybe it would be good to have it clean. Looking at places where z_add_timeout is used, it seems to me that it could take k_timeout_t for legacy code as well. There is only one place where `z_add_timeout` is called with ticks without earlier conversion from milliseconds: `msgdev.c` and it is a sample using internal api so i guess it could be solved somehow (e.g. force new timeout scheme only).

Reply:
I'd really prefer to do that cleanup later.  I really don't think keeping a few 2-5 line blocks of code unchanged under an #if is that much of a burden.  And the intent is that LEGACY goes away in a release or two anyway.

Reply:
can you please add comment in `z_add_timeout` explaining why for legacy implementation value stored in `k_timeout_t` is not actual k_timeout_t but ticks. Someone else looking at that code may get confused (someone else can be me or you in few months :)).